network/src/peers/peer/outbound_handler.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::time::Instant;
+
+use tokio::sync::{mpsc, oneshot};
+
+use crate::{stats, NetworkError, Payload, Peer};
+
+use super::network::PeerIOHandle;
+
+pub(super) enum PeerAction {
+    Disconnect,
+    Send(Payload),
+    Get(oneshot::Sender<Peer>),
+    QualityJudgement,
+    CancelSync,
+    GotSyncBlock,
+    ExpectingSyncBlocks(u32),
+    SoftFail,
+}
+
+#[derive(Clone, Debug)]
+pub struct PeerHandle {
+    pub(super) sender: mpsc::Sender<PeerAction>,
+}
+
+impl PeerHandle {
+    pub async fn load(&self) -> Option<Peer> {
+        metrics::increment_gauge!(stats::QUEUES_OUTBOUND, 1.0);
+        let (sender, receiver) = oneshot::channel();
+        self.sender.send(PeerAction::Get(sender)).await.ok()?;
+        receiver.await.ok()
+    }
+
+    pub async fn judge_bad(&self) {
+        metrics::increment_gauge!(stats::QUEUES_OUTBOUND, 1.0);
+        self.sender.send(PeerAction::QualityJudgement).await.ok();
+    }
+
+    /// returns true if disconnected, false if not connected anymore
+    pub async fn disconnect(&self) -> bool {
+        metrics::increment_gauge!(stats::QUEUES_OUTBOUND, 1.0);
+        self.sender.send(PeerAction::Disconnect).await.is_ok()
+    }
+
+    pub async fn send_payload(&self, payload: Payload) {
```

Comment:
Is there a difference between send_payload and send_message?

Reply:
where do you see `send_message`? this file only has `send_payload`. we don't use `Message` structs because there is no routing or directionality left unclear when this code is called (it's outbound to a specific peer)